Lib/multiprocessing/shared_memory.py

```diff
@@ -1,225 +1,316 @@
-"Provides shared memory for direct access across processes."
+"""Provides shared memory for direct access across processes.
+
+The API of this package is currently provisional. Refer to the
+documentation for details.
+"""
 
 
 __all__ = [ 'SharedMemory', 'PosixSharedMemory', 'WindowsNamedSharedMemory',
-            'ShareableList', 'shareable_wrap',
-            'SharedMemoryServer', 'SharedMemoryManager', 'SharedMemoryTracker' ]
+            'ShareableList',
+            'SharedMemoryServer', 'SharedMemoryManager' ]
 
 
-from functools import reduce
+from functools import partial
 import mmap
-from .managers import DictProxy, SyncManager, Server
+from .managers import dispatch, BaseManager, Server, State, ProcessError
 from . import util
 import os
-import random
 import struct
-import sys
+import secrets
 try:
-    from _posixshmem import _PosixSharedMemory, Error, ExistentialError, O_CREX
+    import  _posixshmem
+    from os import O_CREAT, O_EXCL, O_TRUNC
 except ImportError as ie:
-    if os.name != "nt":
-        # On Windows, posixshmem is not required to be available.
-        raise ie
-    else:
-        _PosixSharedMemory = object
-        class ExistentialError(BaseException): pass
-        class Error(BaseException): pass
-        O_CREX = -1
+    O_CREAT, O_EXCL, O_TRUNC = 64, 128, 512
+
+O_CREX = O_CREAT | O_EXCL
+
+if os.name == "nt":
+    import ctypes
+    from ctypes import wintypes
+
+    kernel32 = ctypes.windll.kernel32
```

Comment:
This should be `kernel32 = ctypes.WinDLL('kernel32', use_last_error=True)`. With `ctypes.windll`, it's using a globally cached `WinDLL` instance. This in turn caches function pointers, which leads to conflicting prototypes between packages. Also, `use_last_error` is required for `ctypes.get_last_error()`, which is more reliable than FFI-based `ctypes.GetLastError()`.

Is this ctypes usage just temporary for experimenting until the API is finalized? Ultimately I hope this uses _winapi instead of ctypes.

Reply:
Thanks, eryksun -- you rock.  Changed both in commit 868b83d681c3e11dd6de22631ac61493a5b73ded.

I believe the API is close enough to being finalized that we can move this to _winapi now.  I would very much appreciate your help reviewing that move, if I can talk you into that.

Reply:
I have moved this over to use _winapi now, changed in commit 9d83b060b6f0e0df82a8f2fe4833495bf1830181.

If you are up for helping review this too, it would again be very much appreciated.